### PR TITLE
ODSC llms integration improvements

### PR DESCRIPTION
**Pull request by mingkang111**
# Description

This PR aims to merge llms integration improvements into feature branch.

# User Experience

### Generic Usage
User will be able to specify the common model parameters: `max_tokens`, `temperature`, `k`, `p` and `best_of`  during initialization. For other parameters, user can pass it through `model_kwargs`. 

```
llm = OCIModelDeploymentLLM(
    endpoint=f"https://modeldeployment.us-ashburn-1.oci.customer-oci.com/{ocid}/predict",
    model="odsc-llm",
    max_tokens=10,
    temperature=0.8,
    k=-1,
    p=0.75,
    best_of=3,
    model_kwargs={"n": 3},  # params that not specified in class
    streaming=True,
)
```
### Framework Specific Usage
User will be able to specify the framework related model parameters during initialization.
```
llm = OCIModelDeploymentVLLM(
    endpoint=f"https://modeldeployment.us-ashburn-1.oci.customer-oci.com/{vllm_md_ocid}/predict",
    model="odsc-llm",
    max_tokens=10,
    temperature=0.8,
    k=-1,
    p=0.75,
    best_of=3,
    n=3,
    frequency_penalty=0.0,
    presence_penalty=0.0,
    use_beam_search=False,
    ignore_eos=False,
    logprobs=None,
    model_kwargs={},  # other params that not specified in class
    streaming=True,
)

llm = OCIModelDeploymentTGI(
    endpoint=f"https://modeldeployment.us-ashburn-1.oci.customer-oci.com/{ocid}/predict",
    model="odsc-llm",
    max_tokens=10,
    temperature=0.8,
    k=-1,
    p=0.75,
    best_of=3,
    do_sample=True,
    watermark=True,
    return_full_text=False,
    streaming=True,
)
```

# Invocation

```
llm.invoke("tell me a joke")
llm.generate(["tell me a joke", "what is deep learning?"])
llm.stream("tell me a joke")
resp = await llm.ainvoke("Write a 20-word article about rivers.")
resp = await llm.agenerate(prompts=["Write a 20-word article about rivers."])
resp = await llm.astream("Write a 20-word article about rivers.")
```

## Customized usage

User can inherit from our plugin and overrwrite the `_process_response`, `_process_stream_response`, `_construct_json_body` for satisfying customized needed. In this case, when user wants to deploy chain, the code for custom class need to be added into model artifact.

For example:

```
class MyModelPlugin(OCIModelDeploymentVLLM):
    def _process_stream_response(self, response_json:dict) -> GenerationChunk:
           print("my customized output stream handler")
           return GenerationChunk()

    def _process_response(self, response_json:dict) -> List[Generation]:
           print("my customized output handler")
           return [Generation()]

    def _construct_json_body(self, prompt: str, param:dict) -> dict:
           print("my customized input handler")
           return {}

llm = MyModelPlugin(
    endpoint=f"https://modeldeployment.us-ashburn-1.oci.customer-oci.com/{ocid}/predict",
    model="odsc-llm",
}

llm.invoke("tell me a joke")
```

### CODE REVIEW

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 1]
```diff
@@ -123,94 +401,128 @@ def _call(
...
+
+    def _parse_stream_line(self, line: bytes) -> Optional[str]:
+        line = line.strip()
+        if line:
+            line = line.decode("utf-8")
+            if "[DONE]" in line:
+                return None
+
+            if line.startswith("data:"):
+                return line[6:] if line.startswith("data: ") else line[5:]
```
Comment (mrDzurb) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: I think it would be more safe to delete the data: and all the whitespaces after it in the line.

Comment (mrDzurb) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: If there is no ```"data:"``` in the line, shouldn't we just return ```line``` as is?

Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: Hi Dmitrii, I am pretty sure I encountered the case that start with `data:` without space, probably when the llamacpp still working. I can try with another deployment to replicate the case. 

When there is no "data:" in the line,  it hit the end of streaming. Yes, we can return line. I am returning None to avoid error from `json.load()`. Let me try again on this case to see if `json.load` will throw error.

Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: Fixed!

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py
Comment (qiuosier) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:1]: Please fix the ruff check failures.

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb
Comment (qiuosier) [inline docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb:1]: Let's add a link to AQUA in the introduction. Maybe something like this:

Users can easily deploy, fine-tune, and evaluate foundation models using the [AI Quick Actions](https://docs.oracle.com/en-us/iaas/data-science/using/ai-quick-actions.htm)

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 2]
```diff
@@ -123,93 +264,293 @@ def _call(
...
+        """Constructs the request body as a dictionary (JSON)."""
+        return {
+            "prompt": prompt,
+            **params,
+        }
+
+    def _invocation_params(self, stop: Optional[List[str]], **kwargs: Any) -> dict:
+        """Combines the invocation parameters with default parameters."""
+        params = self._default_params
+        params["stop"] = stop or params["stop"]
```
Comment (qiuosier) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: `self.stop` has a default of `None`. This will cause `KeyError` if user did not specify `stop`. We should add a default of `None`?
```
params["stop"] = stop or params.get("stop", [])
```

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 3]
```diff
@@ -123,93 +264,293 @@ def _call(
...
-                response.status_code,
-                response.content,
+    def _construct_json_body(self, prompt: str, params: dict) -> dict:
+        """Constructs the request body as a dictionary (JSON)."""
+        return {
+            "prompt": prompt,
+            **params,
+        }
+
+    def _invocation_params(self, stop: Optional[List[str]], **kwargs: Any) -> dict:
```
Comment (qiuosier) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: `stop` is `Optional` but there is no default value at the method level here.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 4]
```diff
@@ -256,33 +599,26 @@ def _llm_type(self) -> str:
...
-            if self.k > 0
-            else None,  # `top_k` must be strictly positive'
+            "top_k": (
+                self.k if self.k > 0 else None
+            ),  # `top_k` must be strictly positive'
             "top_p": self.p,
             "do_sample": self.do_sample,
             "return_full_text": self.return_full_text,
             "watermark": self.watermark,
+            **self.model_kwargs,
```
Comment (qiuosier) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: Need to add `stop` here?
```
"stop": self.stop,
```

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 5]
```diff
@@ -256,33 +599,26 @@ def _llm_type(self) -> str:
     def _default_params(self) -> Dict[str, Any]:
         """Get the default parameters for invoking OCI model deployment TGI endpoint."""
         return {
+            "model": DEFAULT_MODEL_NAME,  # can be any
             "best_of": self.best_of,
             "max_new_tokens": self.max_tokens,
```
Comment (qiuosier) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: Looks like the new TGI API spec also uses `max_tokens` instead of `max_new_tokens`.
https://huggingface.github.io/text-generation-inference/#/Text%20Generation%20Inference/completions

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 6]
```diff
@@ -256,33 +599,26 @@ def _llm_type(self) -> str:
...
             "temperature": self.temperature,
-            "top_k": self.k
-            if self.k > 0
-            else None,  # `top_k` must be strictly positive'
+            "top_k": (
```
Comment (qiuosier) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: Seems like top_k is not in the new TGI spec?

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb
Comment (mingkang111) [reply docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb:1]: Added.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:1]: Fixed

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb [hunk 7]
```diff
@@ -29,23 +29,17 @@
     "## Prerequisite\n",
     "\n",
     "### Deploy model\n",
-    "Check [Oracle GitHub samples repository](https://github.com/oracle-samples/oci-data-science-ai-samples/tree/main/model-deployment/containers/llama2) on how to deploy your llm on OCI Data Science Model deployment.\n",
+    "You can easily deploy, fine-tune, and evaluate foundation models using the [AI Quick Actions](https://docs.oracle.com/en-us/iaas/data-science/using/ai-quick-actions.htm) on OCI Data Science Model deployment. Check [Oracle GitHub samples repository](https://github.com/oracle-samples/oci-data-science-ai-samples/tree/main/model-deployment/containers/llama2) on for more deployment examples.\n",
```
Comment (mrDzurb) [inline docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb]: If we introduce AQUA here, maybe in this case it would be better to point out to the AQUA examples instead of deploying llama2?

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb [hunk 8]
```diff
@@ -6,7 +6,7 @@
    "source": [
     "# OCI Data Science Model Deployment Endpoint\n",
     "\n",
-    "[OCI Data Science](https://docs.oracle.com/en-us/iaas/data-science/using/home.htm) is a fully managed and serverless platform for data science teams to build, train, and manage machine learning models in the Oracle Cloud Infrastructure.\n",
+    "> [OCI Data Science](https://docs.oracle.com/en-us/iaas/data-science/using/home.htm) is a fully managed and serverless platform for data science teams to build, train, and manage machine learning models in the Oracle Cloud Infrastructure.\n",
```
Comment (mrDzurb) [inline docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb]: I think it would be better to remove ```>``` at the beginning of the paragraph.

Comment (mrDzurb) [inline docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb]: NIT: 

```
OCI Data Science is a fully managed, serverless platform that enables data science teams to build, train, and manage machine learning models on Oracle Cloud Infrastructure.

This notebook demonstrates how to utilize an LLM hosted on an OCI Data Science Model Deployment.

For authentication, the `oracle-ads` library is used to automatically load credentials required for invoking the endpoint.
```

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb (hunk 7 above)
Comment (mrDzurb) [inline docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb]: NIT: Maybe we can say: ```For additional deployment examples, please visit the Oracle GitHub samples repository.```

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb (hunk 8 above)
Comment (mrDzurb) [reply docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb]: I would rather suggest to add prerequisites section and put installing ADS and info about policies there.

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb [hunk 9]
```diff
@@ -54,6 +48,34 @@
     "## Example"
```
Comment (mrDzurb) [inline docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb]: NIT: Example -> Examples?

In examples I think we can add more comments about what would be the difference between using ```OCIModelDeploymentLLM``` and ```OCIModelDeploymentVLLM```. I would suggest to add more structured examples:

**Something like:**
```
Examples:

- Model Invocation
   - Generic Invocation: Universal method applicable across different models.
   - Framework-Specific Invocation: Provides examples for specific frameworks.
- Asynchronous Invocation: How to invoke models asynchronously to optimize performance.
- Streaming Results
```

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 4 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: fixed.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 5 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: fixed.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 2 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: fixed.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 3 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: fixed.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 6 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: As we discussed offline, we introducing `endpoint_spec` for incorporate both `/generate` and `/v1/completions` spec.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 10]
```diff
@@ -60,19 +79,320 @@ def validate_environment(  # pylint: disable=no-self-argument
...
             values["auth"] = ads.common.auth.default_signer()
+
         values["endpoint"] = get_from_dict_or_env(
             values,
             "endpoint",
             "OCI_LLM_ENDPOINT",
         )
         return values
 
+    def _headers(self, is_async=False, body=None) -> Dict:
```
Comment (mrDzurb) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:107]: Just curious, will this work?

```
headers = {
        "Content-Type": DEFAULT_CONTENT_TYPE_JSON,
    }

    if is_async:
        signer = self.auth["signer"]
        req = requests.Request("POST", self.endpoint, json=body)
        prepared_req = req.prepare()
        signed_req = signer(prepared_req)
        headers.update(signed_req.headers)

    if self.streaming:
        headers.update({
            "enable-streaming": "true",
            "Accept": "text/event-stream",
        })

    return headers
```

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 11]
```diff
@@ -60,19 +79,320 @@ def validate_environment(  # pylint: disable=no-self-argument
...
+                    **kwargs,
+                )
+                self._check_response(response)
+                return response
+            except TokenExpiredError as e:
+                raise e
+            except Exception as err:
+                logger.debug(
+                    f"Requests payload: {data}. Requests arguments: "
+                    f"url={self.endpoint},timeout={request_timeout},stream={stream}."
```
Comment (mrDzurb) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: NIT: space at the end?

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 12]
```diff
@@ -60,19 +79,320 @@ def validate_environment(  # pylint: disable=no-self-argument
...
+                    f"Additional request kwargs={kwargs}."
+                )
+                raise ValueError(
```
Comment (mrDzurb) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: Maybe instead of using ```ValueError``` we can create a custom error?

```
class InferenceEndpointError(Exception):
    """Custom exception for errors occurring at the inference endpoint."""
    def __init__(self, message: str, original_exception: Exception) -> None:
        super().__init__(message)
        self.original_exception = original_exception
```


```
raise InferenceEndpointError(
                f"Error occurred at inference endpoint: {str(err)}", err
            ) from err
```

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 13]
```diff
@@ -60,19 +79,320 @@ def validate_environment(  # pylint: disable=no-self-argument
...
+        """
+        try:
+            response.raise_for_status()
+        except requests.exceptions.HTTPError as http_err:
+            status_code = (
+                response.status_code
+                if hasattr(response, "status_code")
+                else response.status
+            )
+            if status_code == 401 and self._refresh_signer():
```
Comment (mrDzurb) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:254]: Can you elaborate on this logic please?

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 14]
```diff
@@ -60,19 +79,320 @@ def validate_environment(  # pylint: disable=no-self-argument
...
+                raise TokenExpiredError() from http_err
+            else:
```
Comment (mrDzurb) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: NIT: ```else``` is not required?

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 15]
```diff
@@ -60,19 +79,320 @@ def validate_environment(  # pylint: disable=no-self-argument
...
+            Optional[str]:
+                The processed line as a string if valid, otherwise `None`.
+        """
+        line = line.strip()
+        if line:
+            line = line.decode("utf-8")
+            if "[DONE]" in line:
+                return None
+
+            if line.startswith("data:"):
```
Comment (mrDzurb) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: NIT: ```line.lower().startswith("data:"):```?

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb (hunk 7 above)
Comment (qiuosier) [reply docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb]: yes, the llama2 is outdated :)

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 16]
```diff
@@ -60,19 +89,325 @@ def validate_environment(  # pylint: disable=no-self-argument
...
```
Comment (qiuosier) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: When the line does not start with `data:`, does it mean something unexpected is happening?
In that case, we simply discard the line? or shall we do the following?
```
if line.startswith("data:"):
    return line[5:].lstrip()
return line
```

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 17]
```diff
@@ -60,19 +89,325 @@ def validate_environment(  # pylint: disable=no-self-argument
...
+    ) -> Any:
+        """Use tenacity to retry the completion call."""
+        retry_decorator = _create_retry_decorator(self, run_manager=run_manager)
+
+        @retry_decorator
+        def _completion_with_retry(**kwargs: Any) -> Any:
+            try:
+                request_timeout = kwargs.pop("request_timeout", DEFAULT_TIME_OUT)
+                data = kwargs.pop("data")
+                stream = kwargs.pop("stream", self.streaming)
```
Comment (qiuosier) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:157]: For `TGI`, I have to add `data["stream"] = stream` in order to enable the streaming when `stream` is `True`. Looks like the OpenAI spec also mentioned this: https://platform.openai.com/docs/api-reference/completions/create#completions-create-stream

Does it work without `stream` in data for vllm?

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 18]
```diff
@@ -60,19 +89,325 @@ def validate_environment(  # pylint: disable=no-self-argument
...
+        """Use tenacity to retry the async completion call."""
+        retry_decorator = _create_retry_decorator(self, run_manager=run_manager)
+
+        @retry_decorator
+        async def _completion_with_retry(**kwargs: Any) -> Any:
+            try:
+                request_timeout = kwargs.pop("request_timeout", DEFAULT_TIME_OUT)
+                data = kwargs.pop("data")
+                stream = kwargs.pop("stream", self.streaming)
+
```
Comment (qiuosier) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:199]: I have to added
```
data["stream"] = stream
```
for streaming to work.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 19]
```diff
@@ -82,36 +417,99 @@ def _identifying_params(self) -> Dict[str, Any]:
...
+        Example:
+            .. code-block:: python
 
-    def _call(
+                response = await oci_md.ainvoke("Tell me a joke.")
+                response = await oci_md.agenerate(["Tell me a joke."])
+        """  # noqa: E501
+        generations: List[List[Generation]] = []
+        params = self._invocation_params(stop, **kwargs)
+        for prompt in prompts:
```
Comment (qiuosier) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:536]: Maybe it is better to check if `prompts` is a list here so that user will not pass in a single string by mistake (which will cause the agenerate to run for a long time?
```
prompts = [prompts] if isinstance(prompts, str) else prompts
```

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 13 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:254]: sure, this is refactoring this part: https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py#L181-L191 . Allowing retry for refresh token before invoking.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 12 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: I used `RuntimeError` at beginning, then switch back to `ValueError` as I have observed that all other provider using this name.  If `ValueError` is not ok, then probably `RuntimeError` can be a better name.

Comment (mrDzurb) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: I'm ok with either. However, using ```ValueError``` for general exceptions might not be the best approach, especially if the error does not specifically relate to an invalid value.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 18 above)
Comment (mrDzurb) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:199]: +1

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 16 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: During the test, one case of line not start with `data:` is at the ending line, it gave an empty line, same as `[Done]` case.  Returning `None` here is to skip this line which can avoid failure in `_process_stream_response` downstream, and not breaking the entire streaming.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 18 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:199]: I added into default_params for vllm and tgi to fix this issues.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 19 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:536]: added .

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 17 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:157]: I added `stream` into default params for vllm and tgi subclass to fix this issue.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 15 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: fixed.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 14 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: fixed

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 11 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: fixed

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb (hunk 7 above)
Comment (mingkang111) [reply docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb]: fixed

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb (hunk 8 above)
Comment (mingkang111) [reply docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb]: fixed

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb (hunk 9 above)
Comment (mingkang111) [reply docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb]: > add more structured examples:

I did this in `ChatModel` but not in here as LC advertise to Chatmodel not text completions in many providers. In addition, since LangChain already have many invocation/async/streaming examples, our doc can keep it simple. User are expected to learn these langchain's functionalities from LangChain's doc. 

In case of difference between using OCIModelDeploymentLLM and OCIModelDeploymentVLLM, I put into the inline comments in code. I haven't put it into details as I need to know which one we want to advertise more. I would prefer we have discussion on documentation's content in a meeting later.

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb (hunk 8 above)
Comment (mingkang111) [reply docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb]: Hi Dmitrii, I have removed `>`.

There is a cell for installing ads. The policies would be better to learn from oracle's doc and ads doc.

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb (hunk 7 above)
Comment (mingkang111) [reply docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb]: Updated the link to deploy with smc example.

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 12 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: Sure. Changed to `RuntimeError`.

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb (hunk 9 above)
Comment (mingkang111) [reply docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb]: I have provided the link for api doc regarding differences between framework specific initialization and generic initialization. I believe in this way can provide the most valuable information for user.

**Review by qiuosier — APPROVED** [review]

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 20]
```diff
@@ -60,19 +93,369 @@ def validate_environment(  # pylint: disable=no-self-argument
...
+                    data=data,
+                    timeout=request_timeout,
+                    stream=stream,
+                    **kwargs,
+                )
+                self._check_response(response)
+                return response
+            except TokenExpiredError as e:
+                raise e
+            except Exception as err:
```
Comment (mrDzurb) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:173]: This will not show us real error form the endpoint? For example if i pass `top_p=-1`, will the error show me that top_p is incorrect?

**Review by mrDzurb — APPROVED** [review]

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 20 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:173]: I believe it will show. When parameter is incorrect, the server won't return 200 status code and thus the method `self._check_response(response)` will raise the error, which will show real error form the endpoint. 

e.g.
```
RuntimeError: Error occurs by inference endpoint: Server error: 404 Client Error: Not Found for url: https://modeldeployment.us-ashburn-1.oci.customer-oci.com/<>/predict. 
Message: {"object":"error","message":"The model `llm` does not exist.","type":"invalid_request_error","param":null,"code":null}
```
